Message handler: add quote_begin, quote_end commands

tautschnig opened the pull request:
These will permit custom formatting of quoting for each message handler, and for a start introduce <quote> tags for XML. See #4875 for discussion.

Please review commit-by-commit.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Message handler: add quote_begin, quote_end commands

These will permit custom formatting of quoting for each message handler,
See #4875 for discussion.

tautschnig committed:
Message handler: add quote_begin, quote_end commands

These will permit custom formatting of quoting for each message handler,
See #4875 for discussion.

tautschnig committed:
Replace use of ` in messages by quote_begin (and ' by quote_end)

This continues the work of #4875, but now using quote_begin/quote_end
stream modifiers.

tautschnig committed:
Replace use of ` in messages by quote_begin (and ' by quote_end)

This continues the work of #4875, but now using quote_begin/quote_end
stream modifiers.

tautschnig committed:
Replace quoting via " in messages by quote_begin and quote_end

This is in the spirit of #4875, but now using quote_begin/quote_end stream
modifiers.

tautschnig committed:
Replace quoting via " in messages by quote_begin and quote_end

This is in the spirit of #4875, but now using quote_begin/quote_end stream
modifiers.

martin-cs reviewed (approved):
Seems good.  It would be nice if this also had the implementation of this for XML so that it was clear why it was valueable and no-one will accidentally 'simplify' it out.

peterschrammel commented on src/util/ui_message.h:
```diff
@@ -94,10 +94,75 @@ class ui_message_handlert : public message_handlert
+      case 96: // bright_cyan
+        return std::string();
+      case '<': // quote_begin
+        return "<quote>";
```
Not sure this is a good idea as it changes the XML output format. I'd stick to single quotes for now and introduce a different format in a separate PR if desired.

tautschnig replied:
I have now factored this into a commit of its own, which could thus easily be dropped from the series.

peterschrammel replied:
I would really prefer to drop this commit. IMO, it causes unnecessary hassle to XML output users.

tautschnig replied:
Removed, just left behind a comment that this could be done.

tautschnig committed:
ui_message: centralise SGR-style command suppression

The XML_UI and JSON_UI cases in ui_message_handlert::command()
duplicated the same 17-entry switch/fallthrough list of SGR codes
to suppress. This meant any future addition or removal of a
formatting command had to be made in two places.

Extract a static helper is_sgr_style_command(unsigned c) that
returns true for the SGR parameter values used by messaget (0..4
for reset/bold/faint/italic/underline, 31..36 and 91..96 for
foreground colours and their bright variants). The two duplicated
lists are replaced by a single early-return guarded by the helper,
keeping the quote_begin / quote_end handling per-UI as before.

Behaviour-preserving: the same set of codes is suppressed on the
same UIs; the quote commands still render as single-quote on all
three UIs. Verified by comparing PLAIN, XML, and JSON output of a
diagnostic that exercises both quoting and colour before and after
this change.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig commented on src/util/ui_message.h:
```diff
@@ -96,10 +96,80 @@ class ui_message_handlert : public message_handlert
+    default:
+      return false;
+    }
   }
```
Two minor maintainability points on the new helper:

1. The body is large enough to be worth moving to `ui_message.cpp`. Today it is inlined into every TU that includes `ui_message.h`, which is a lot. Same comment applies to `command()` above.
2. There is no unit test exercising this helper, even though the whole point of the refactor was to make it easier to add/remove SGR codes. A quick `TEST_CASE("is_sgr_style_command covers all messaget styling codes", "[util][ui_message]")` that walks the `messaget::commandt` constants and asserts the predicate matches (and conversely that `quote_begin`/`quote_end` do *not* match) would lock the contract down. It would also catch the case where the doc-string range disagrees with the `case` list (they currently agree, but they're maintained independently).

tautschnig commented on src/util/message.cpp:
```diff
@@ -95,6 +95,8 @@ const messaget::commandt messaget::bright_yellow(93);
 const messaget::commandt messaget::bright_magenta(95);
 const messaget::commandt messaget::bright_cyan(96);
+const messaget::commandt messaget::quote_begin('<');
+const messaget::commandt messaget::quote_end('>');
```
Encoding-choice nit: `'<'` and `'>'` happen to be ASCII 60 and 62, which are valid ECMA-48 SGR parameter slots reserved for ideogram-related styling (60–65). Nothing currently uses them, but if anyone ever adds ideogram modes via this same mechanism we'll have a runtime clash that's easy to miss because it produces no compile-time signal.

A cleaner alternative is to give non-SGR commands values outside the SGR range, e.g. `static const commandt quote_begin(256);`. The `console_message_handlert::command` and `ui_message_handlert::command` already special-case these by value, so widening is essentially free. At minimum, please add a comment here explaining the overload so future readers don't accidentally pick another colliding code.
